### ISSUE TITLE
classlib: Float.scaleneg produces incorrect results

### ISSUE THREAD

**Issue by mossheim** [supercollider/supercollider#3708]
```supercollider
scaleneg(-1, 2) // -2

scaleneg(-1.0, 2.0) // 2.0
```

AFAICT this only affects `Float.scaleneg` - not integer or signal or the ugen. Due to a bad implementation here:

https://github.com/supercollider/supercollider/blob/cca12ff02a774a9ea212e8883551d3565bb24a6f/include/plugin_interface/SC_InlineBinaryOp.h#L556-L568

apparently from the initial commit.

This is clearly an error and I think it should be patched directly. The other option is to deprecate and provide a new function, but:

1. that will take a lot of extra work because this function exists on many classes
2. it will require two sets of unit tests to ensure the old behavior is correctly incorrect
3. either a new primitive will have to be written or scaleneg will be implemented in SC which means a performance hit
4. forcing existing uses of correct versions of scaleneg to migrate seems just as unfriendly as breaking existing uses

**Comment by brianlheim** [comment]
I pushed a branch on this repo at `topic/scaleneg` which patches it in-place and adds tests.

**Comment by patrickdupuis** [comment]
I'm trying this branch out. All the test in TestFloat that make use of `0/0` fail. The result of `scaleneg(-1.0, 0/0).isNaN`, for instance, is `false`. The reason is a bit strange; `0/0` results in `-nan` not `nan`.

**Comment by brianlheim** [comment]
Right. Locally I'm building with Clang so the tests are fine here (it just uses the naive/simple implementation). I should be able to work out the kinks in compiler explorer.

**Comment by brianlheim** [comment]
@snappizz - I'm no longer working on this actively, it is not high priority for me right now.

**Comment by tsmtzs** [comment]
In `include/plugin_interface/SC_InlineBinaryOp.h` the function `scaleneg` is defined as (line `557`)
```supercollider
0.5 * (|a| - a) * (b + 1) + a
```
which gives `a` for `a >= 0` and `- a * b` for `a < 0`.

Thus,  `scaleneg(-1.0, 2.0)` returns `-2.0`. The  example `scaleneg(-1, 2)` is handled with the template at line `548` of the above file which gives the correct value `-2`.

Instead, one can use the formula
```supercollider
min(a, 0) * b + max(a, 0)
```

Since `min(a, 0) = 0.5 * ( a - |a|)`  and `max(a, 0) = 0.5 * (a + |a|)`, the above formula reduces to
```supercollider
0.5 * ((a - |a|) * b + a + |a|)
```

So, we could define `scaleneg` as
https://github.com/tsmtzs/supercollider/blob/188e922d55631e075cab5398dc33c97445eae306/include/plugin_interface/SC_InlineBinaryOp.h#L497-L505

_Remark_: Although I 've tested the code, and added a (passing) unit test in `testsuite/classlibrary/TestSimpleNumber.sc`, I have no serious knowledge of `C++`. This means that the above code might not be appropriate.
